usr/src/uts/common/fs/zfs/metaslab.c

```diff
@@ -2403,6 +2478,9 @@ metaslab_sync_done(metaslab_t *msp, uint64_t txg)
 	 * defer_tree -- this is safe to do because we've just emptied out
 	 * the defer_tree.
 	 */
+	if (spa_get_auto_trim(spa) == SPA_AUTO_TRIM_ON &&
+	    !vd->vdev_man_trimming)
+		range_tree_walk(*defer_tree, metaslab_trim_add, msp);
```

Comment:
The idea here was to make correctness checking a bit simpler. If there is some space in the current trimset, then it absolutely has to be in ms_tree as well. The couple txgs that the space sits in the defer tree doesn't really bother us.
The defer_allowed case is actually an oversight. I'll fix that. We want to trim everything just as it's entering ms_tree.